scripts/ubuntu-20-provision.sh

```diff
@@ -37,163 +38,206 @@ function install_package_version() {
 }
 
 ## Copy custom scripts
-cp /tmp/add_auth_key_to_user.sh /usr/local/bin/add_auth_key_to_user.sh
-chmod a+x /usr/local/bin/add_auth_key_to_user.sh
+function copy_custom_scripts() {
+  cp /tmp/add_auth_key_to_user.sh /usr/local/bin/add_auth_key_to_user.sh
+  chmod a+x /usr/local/bin/add_auth_key_to_user.sh
+}
 
-## Disable and Remove Unattended APT Upgrades
-echo 'APT::Periodic::Enable "0";' > /etc/apt/apt.conf.d/10cloudinit-disable
-apt-get purge -y unattended-upgrades || true # Do not fail if the package does not exist
+## All the clean for apt
+function clean_apt() {
+  ## Disable and Remove Unattended APT Upgrades
+  echo 'APT::Periodic::Enable "0";' > /etc/apt/apt.conf.d/10cloudinit-disable
+  apt-get purge -y unattended-upgrades || true # Do not fail if the package does not exist
 
-## Remove unused packages
-apt-get purge -y snap lxd || true # Do not fail if the package does not exist
-apt-get autoremove --purge -y
+  ## Remove unused packages
+  apt-get purge -y snap lxd || true # Do not fail if the package does not exist
+  apt-get autoremove --purge -y
 
-## Ensure the machine is up-to-date
-apt-get update
-apt-get upgrade -y
+  ## Ensure the machine is up-to-date
+  apt-get update
+  apt-get upgrade -y
+}
 
 ## Ensure Docker is installed as per https://docs.docker.com/engine/install/ubuntu/
-apt-get install -y --no-install-recommends \
-  ca-certificates \
-  curl \
-  gnupg \
-  lsb-release \
-  gnupg-agent \
-  software-properties-common
-
-curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
-echo \
-  "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
-  $(lsb_release -cs) stable" | tee /etc/apt/sources.list.d/docker.list > /dev/null
-apt-get update
-apt-get install -y --no-install-recommends docker-ce
+function install_docker() {
+  apt-get install -y --no-install-recommends \
+    ca-certificates \
+    curl \
+    gnupg \
+    lsb-release \
+    gnupg-agent \
+    software-properties-common
+
+  curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
+  echo \
+    "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
+    $(lsb_release -cs) stable" | tee /etc/apt/sources.list.d/docker.list > /dev/null
+  apt-get update
+  apt-get install -y --no-install-recommends docker-ce
+}
 
 ## Ensure that the Jenkins Agent commons requirements are installed
-apt-get install -y --no-install-recommends \
-  make \
-  unzip \
-  zip \
-  jq \
-  parallel
+function install_JA_requirements(){
+  apt-get install -y --no-install-recommends \
+    make \
+    unzip \
+    zip \
+    jq \
+    parallel
+}
 
 # setup qemu
```

Comment:
```suggestion
## setup qemu
```
Not sure why the use of `#` or `##`

Reply:
I tend to use double by default, to allow "simple" for sub sections.